test/e2e/storage/testsuites/provisioning.go

```diff
@@ -527,6 +528,266 @@ func (p *provisioningTestSuite) DefineTests(driver storageframework.TestDriver,
+		biggerSize := originalSize.DeepCopy()
+		biggerSize.Add(resource.MustParse("1Gi"))
+
+		biggerROClaim := l.pvc.DeepCopy()
+		biggerROClaim.Name = "bigger-claim"
+		biggerROClaim.Spec.Resources.Requests = v1.ResourceList{
+			v1.ResourceStorage: biggerSize,
+		}
+		biggerROClaim.Spec.AccessModes = []v1.PersistentVolumeAccessMode{v1.PersistentVolumeAccessMode(v1.ReadOnlyMany)}
+
+		// Create readonlymany claim w/ same size as original, then test to make
+		// sure it has expected content when restored.
+		sameSizeROClaim := l.pvc.DeepCopy()
+		sameSizeROClaim.Name = "same-size-claim"
+		sameSizeROClaim.Spec.AccessModes = []v1.PersistentVolumeAccessMode{
+			v1.PersistentVolumeAccessMode(v1.ReadOnlyMany),
+		}
+
+		// For use in TestDynamicProvisioning - test expected content exists in
+		// restored volume
+		l.testCase.PvCheck = func(ctx context.Context, claim *v1.PersistentVolumeClaim) {
+			ginkgo.By("checking whether the created volume has the pre-populated data")
+			tests := []e2evolume.Test{
+				{
+					Volume:          *storageutils.CreateVolumeSource(claim.Name, false /* readOnly */),
+					Mode:            pattern.VolMode,
+					File:            "index.html",
+					ExpectedContent: expectedContent,
+				},
+			}
+			e2evolume.TestVolumeClientSlow(ctx, f, testConfig, nil, "", tests)
+		}
+
+		for _, claim := range []*v1.PersistentVolumeClaim{sameSizeROClaim, biggerROClaim} {
+			l.testCase.Claim = claim
+			l.testCase.TestDynamicProvisioning(ctx)
+
+			// Do not check fs size for block volumes
+			if pattern.VolMode == "Block" {
+				continue
+			}
+
+			// Create pvc + pod to get mount path of fs
+			originalClaim, err = l.testCase.Client.CoreV1().PersistentVolumeClaims(claim.Namespace).Create(ctx, claim, metav1.CreateOptions{})
+			framework.ExpectNoError(err, "Failed to create pvc: %v", err)
+			createdClaims := []*v1.PersistentVolumeClaim{originalClaim}
+			var podConfig *e2epod.Config = &e2epod.Config{
+				NS:            f.Namespace.Name,
+				PVCs:          createdClaims,
+				NodeSelection: testConfig.ClientNodeSelection,
+				PVCsReadOnly:  true,
+			}
+			pod, err := e2epod.CreateSecPod(ctx, l.testCase.Client, podConfig, l.testCase.Timeouts.DataSourceProvision)
+			framework.ExpectNoError(err, "Failed to create pod: %v", err)
+			mountpath := findVolumeMountPath(pod, originalClaim)
+			gomega.Expect(mountpath).ShouldNot(gomega.BeEmpty())
+
+			// Get actual size of the restored filesystem.
+			restoredFSSize, err := getFilesystemSizeBytes(pod, mountpath)
+			framework.ExpectNoError(err, "Failed to obtain filesystem size of a volume mount: %v", err)
+
+			msg := fmt.Sprintf("Filesystem resize occurred even though it is readonly. "+
+				"Restored fs size: %v bytes is not the same as origin fs size: %v bytes.\n"+
+				"HINT: Your driver needs to check the volume and skip resize if it's readonly.",
+				restoredFSSize, originFSSize)
+			gomega.Expect(restoredFSSize).Should(gomega.BeEquivalentTo(originFSSize), msg)
```

Comment:
I am not sure if we can require equivalence in capacity even if it's ROX. I think NFS-like volume types can support a larger capacity despite being read-only. I think the proper behavior for plugins that can't support the larger size is to fail the CreateVolume call.

@saikat-royc @gnufied  wdyt?

Reply:
Yep. This seems like a problem in case nfs volumes are being provisioned off same directory or something.

Reply:
Thanks for pointing this out. The main goal here was just to make sure that a resize didn't occur when it shouldn't - maybe we add a anti-capability here and only opt-in drivers that we believe should have the volume be the same size before and after (PD, filestore, etc?)

Reply:
I think we can leverage the resizeFromRestore anti capability: https://github.com/kubernetes/kubernetes/pull/114628/files#diff-d7be57591f195129598536ab483c8e93bc048d836c09ed47b409e6e6b5579bffR189

In the case of PD driver, it can't support resize from restore for ROX.

Reply:
PD driver does support general resize from restore as of https://github.com/kubernetes-sigs/gcp-compute-persistent-disk-csi-driver/pull/972 and that's what that as per https://github.com/kubernetes/kubernetes/blob/master/test/e2e/storage/drivers/csi.go#L829 that anti-capability will soon be removed from PD. 

I think we need an anti-capability here that's specific to Resize+ROX. I'll add CapFSROXResizeFromSourceNotSupported.

Reply:
Alrighty added CapFSROXResizeFromSourceNotSupported to this test, but down below right before we check filesystem size. Before that, this test also checks that the clone occurred properly, which I think nfs, etc would still benefit from.